R/sharepoint_client.R

```diff
@@ -0,0 +1,136 @@
+    #'
+    #' This manages authenticating with sharepoint via sending credentails to
+    #' microsoft to retrieve an access token which it then sends to sharepoint
+    #' to retrieve cookies used for subsequent authentication.
+    #'
+    #' @param sharepoint_url Root URL of sharepoint site to download from
+    #' @return A new `sharepoint_client` object
+    initialize = function(sharepoint_url) {
+      self$sharepoint_url <- sharepoint_url
+      private$handle <- httr::handle(sharepoint_url)
+
+      creds <- get_credentials()
+      response <- httr::POST(
+        "https://login.microsoftonline.com/extSTS.srf",
+        body = prepare_security_token_payload(self$sharepoint_url, creds))
+      ## Not sure if this ever returns a non 200 response but left here to
+      ## be safe. On failed auth it sends a different set of xml but still 200
+      if (response$status_code != 200) {
+        stop(sprintf("Failed to authenticate user '%s'.", creds$username))
+      }
+      ## Note that httr preserves cookies and settings over multiple requests
+      ## via the handle. Means that if we auth once and retrieve cookies httr
+      ## will automatically send these on subsequent requests if using the same
+      ## handle object
+      security_token <- parse_security_token_response(response)
+      if (is.na(security_token)) {
+        stop(sprintf("Failed to retrieve security token for user '%s'.",
+                     creds$username))
+      }
+      res <- self$POST("_forms/default.aspx?wa=wsignin1.0",
+                       body = security_token)
+      validate_cookies(res)
+    },
+
+    #' @description
+    #' Send GET request to sharepoint
+    #'
+    #' @param ... Args passed on to httr
+    #'
+    #' @return HTTP response
+    GET = function(...) {
+      self$request(httr::GET, ...)
+    },
+
+    #' @description
+    #' Send POST request to sharepoint
+    #'
+    #' @param ... Args passed on to httr
+    #'
+    #' @return HTTP response
+    POST = function(...) {
+      self$request(httr::POST, ...)
+    },
+
+    #' @description
+    #' Send POST request to sharepoint
+    #'
+    #' @param verb A httr function for type of request to send e.g. httr::GET
+    #' @param path Request path
+    #' @param ... Additional args passed on to httr
+    #'
+    #' @return HTTP response
+    request = function(verb, path, ...) {
+      url <- paste(self$sharepoint_url, path, sep = "/")
+      verb(url, ..., handle = private$handle)
+    }
+  ),
+
+  private = list(
+    handle = NULL
+  )
+)
+
+#' Prepare payload for retrieving security token
+#'
+#' @param url URL for site you are requesting a token for
+#' @param credentials Username and password for site request token
+#'
+#' @return Formatted xml body for security token request
+#' @keywords internal
+prepare_security_token_payload <- function(url, credentials) {
+  payload <- paste(readLines(system.file("security_token_request.xml",
```

Comment:
Can you use the `system_file` pattern we have elsewhere that uses `mustWork = TRUE`, or even a `pointr_file` that does the `package` arg too if you're using more than one inst file

Reply:
Fixed that